### ISSUE TITLE
Tests are failing since SAPUI5 version 1.104

### ISSUE THREAD

**Issue by soraY27** [ui5-community/wdi5#364]
**Describe the bug**
We have raised the version of SAPUI5 in our repository.
As of version 1.104, the tests no longer seem to run.

**To Reproduce**

1. Start some WDI5 tests
2. Change the Version of UI5 to 1.104 or higher (till 1.108)
3. Start the WDI5 tests again
4. Get javascript error: Cannot read properties of undefined (reading 'waitForUI5Options')

**Expected behavior**
It is expected that the tests will go through normally as before, after the version is raised higher than 1.103.

**Logs/Console Output**

```shell
[chrome 107.0.5304.87 mac os x #0-0] javascript error: Cannot read properties of undefined (reading 'waitForUI5Options')
JavaScript stack:
TypeError: Cannot read properties of undefined (reading 'waitForUI5Options')
    at eval (eval at executeAsyncScript (:552:26), <anonymous>:4:65)
    at eval (eval at executeAsyncScript (:552:26), <anonymous>:33:8)
    at eval (eval at executeAsyncScript (:552:26), <anonymous>:33:33)
    at executeAsyncScript (<anonymous>:552:47)
    at apply.forceSelect (<anonymous>:567:29)
    at callFunction (<anonymous>:446:22)
    at <anonymous>:460:23
    at <anonymous>:461:3
  (Session info: headless chrome=107.0.5304.87)
[chrome 107.0.5304.87 mac os x #0-0] javascript error: javascript error: Cannot read properties of undefined (reading 'waitForUI5Options')
[chrome 107.0.5304.87 mac os x #0-0] JavaScript stack:
[chrome 107.0.5304.87 mac os x #0-0] TypeError: Cannot read properties of undefined (reading 'waitForUI5Options')
[chrome 107.0.5304.87 mac os x #0-0]     at eval (eval at executeAsyncScript (:552:26), <anonymous>:4:65)
[chrome 107.0.5304.87 mac os x #0-0]     at eval (eval at executeAsyncScript (:552:26), <anonymous>:33:8)
[chrome 107.0.5304.87 mac os x #0-0]     at eval (eval at executeAsyncScript (:552:26), <anonymous>:33:33)
[chrome 107.0.5304.87 mac os x #0-0]     at executeAsyncScript (<anonymous>:552:47)
[chrome 107.0.5304.87 mac os x #0-0]     at apply.forceSelect (<anonymous>:567:29)
[chrome 107.0.5304.87 mac os x #0-0]     at callFunction (<anonymous>:446:22)
[chrome 107.0.5304.87 mac os x #0-0]     at <anonymous>:460:23
[chrome 107.0.5304.87 mac os x #0-0]     at <anonymous>:461:3
[chrome 107.0.5304.87 mac os x #0-0]   (Session info: headless chrome=107.0.5304.87)
[chrome 107.0.5304.87 mac os x #0-0]     at async clientSide_getControl (/Users/metehan.dogan/U-NIQ/project-X/node_modules/wdio-ui5-service/client-side-js/getControl.js:3:12)
[chrome 107.0.5304.87 mac os x #0-0]
```

**Screenshots**
![image](https://user-images.githubusercontent.com/28558543/200261083-0618a554-8cfa-4b06-be3b-7f9b7c1a10ca.png)

**Runtime Env:**

- `wdi5/wdio-ui5-service`-version: 0.9.13
- `UI5` version: 1.104 and higher (till 1.108)
- `wdio`-version: 7.25.1
- `node`-version: v14.18.3
- OS: macOS Monterey 12.6 
- Browser + Version: 107.0.5304.87

**Comment by dominikfeininger** [comment]
Hi @soraY27,

can you please run the tests in with verbose logging `logLevel: "verbose"` in `wdio.conf.js`  and look for further information.

Thank you

**Comment by soraY27** [comment]
Hi @dominikfeininger 
does that help you?
![image](https://user-images.githubusercontent.com/28558543/200291153-f3e33f85-ce40-4495-8b49-718160439d3b.png)

**Comment by dominikfeininger** [comment]
unfortunately not.

Can you please provide your `wdio.conf.js`

Do you use `skipInjectUI5OnStart: true`

**Comment by soraY27** [comment]
We found out earlier that it is most likely due to BASIC authentication.
Our tests start on the Fiori Launchpad (FLP) and the first test of an app is to click on the app's tile.
When navigating to the app, the login data is lost.

We try to resolve this issue on our end first.
Thank you for the quick feedback so far.

**Comment by vobu** [comment]
pinging @hmanchev - were there changes in the `RecordReplay` api in UI5 > 1.104?

**Comment by vobu** [comment]
i just took a little test-drive w/ [our sample UI5 JS app](https://github.com/ui5-community/wdi5/tree/main/examples/ui5-js-app).
And my hunch is that you ran into a bootstrap URL issue:
with UI5 1.104 [not being an LTS](https://sapui5.hana.ondemand.com/versionoverview.html), the [evergreen bootstrap](https://blogs.sap.com/2022/04/14/sapui5-patch-level-independent-bootstrap/) doesn't apply. 
So when your app bootstraps à la 
  `https://ui5.sap.com/1.104/resources/sap-ui-core.js`
→ app won't work as the minor version is not specified.
`https://ui5.sap.com/1.104.1/resources/sap-ui-core.js` 
→ will work
So might not be related to `wdi5` itself but rather to the bootstrapped UI5 version.
Can you give specifying the minor version a shot please?

**Comment by XSNAT** [comment]
Changing the bootstrap URL didn't help resolve the issue. That is because we use a patch-level independent bootstrap in our project.
![image](https://user-images.githubusercontent.com/113106076/200328963-81c4b8cd-f3e6-4801-8422-bdaeb88ca9aa.png)

**Comment by vobu** [comment]
where do you change the UI5 version then? FLP?

**Comment by MichaelSchmid-AGI** [comment]
So i looked into the Problem from my colleague @XSNAT .
I tested with the following versions: 1.98.0, 1.04.0 ,1.04.1 and 1.08.0, The problem occurs in all of these 4 Versions .

i did this, using the Manifest.json
```
  "sap.ui5": {
    "dependencies": {
      "minUI5Version": "1.108.0",
      "libs": {
        "sap.m": {},
        "sap.ui.core": {},
        "sap.f": {},
        "sap.suite.ui.generic.template": {},
        "sap.ui.comp": {},
        "sap.ui.generic.app": {},
        "sap.ui.table": {},
        "sap.ushell": {}
      }
    },
```
The log of the attempt with 1.08.0:
![image](https://user-images.githubusercontent.com/111853408/200337772-fb577f88-f29f-4a17-b30d-99855c613c78.png)

**Comment by vobu** [comment]
and just to doublecheck: the app itself works with the changed UI5 version(s)?
and you have the `ui5` service configured in `wdio.conf.(j|t)s` (https://ui5-community.github.io/wdi5/#/installation?id=c-manual-installation)? 
Because when the tests error out with all specified UI5 versions, a possible cause could be that `wdi5` itself (via its' technical name `wdio-ui5-service`) is not loaded at test-time. (Plus I just checked that our sample JS app works with some of your quoted UI5 versions)

**Comment by the-docta** [comment]
@vobu our problem (original from this ticket) was with any version starting with 1.104.0 (this was just the indicator, where the problem appeared for the first time), same issue with every version up and including 1.108.0 (bootstrapping with exact verison) // same problem didn't appear in 1.103.0 

in our case, the problem was with the baseUrl being http://user:pass@localhost:4006/... => and clicking the tile on FLP now triggered a full page load (without basic auth part in URL) which ended in basic auth popup on the targeted app, and the appropriate weitForUI5Options error message.  (assuming that basic auth popup blocked some of the processing, but this is just a guess) 

Basically, we are in the process to build up a poor-mans-login-page (use basic auth to generate a session and go to FLP sandbox page after after that)

**Comment by vobu** [comment]
> @vobu our problem (original from this ticket) was with any version starting with 1.104.0 (this was just the indicator, where the problem appeared for the first time), same issue with every version up and including 1.108.0 (bootstrapping with exact verison) // same problem didn't appear in 1.103.0
> 
> in our case, the problem was with the baseUrl being http://user:pass@localhost:4006/... => and clicking the tile on FLP now triggered a full page load (without basic auth part in URL) which ended in basic auth popup on the targeted app, and the appropriate weitForUI5Options error message. (assuming that basic auth popup blocked some of the processing, but this is just a guess)
> 
> Basically, we are in the process to build up a poor-mans-login-page (use basic auth to generate a session and go to FLP sandbox page after after that)

roger that :)
you might also want to track our progress with the configuration-based authentication feature in https://github.com/ui5-community/wdi5/tree/feat/authenticator - there'll be a release next week with it (inclusing docs) I guestimate

**Comment by MichaelSchmid-AGI** [comment]
> and just to doublecheck: the app itself works with the changed UI5 version(s)? and you have the `ui5` service configured in `wdio.conf.(j|t)s` (https://ui5-community.github.io/wdi5/#/installation?id=c-manual-installation)? Because when the tests error out with all specified UI5 versions, a possible cause could be that `wdi5` itself (via its' technical name `wdio-ui5-service`) is not loaded at test-time. (Plus I just checked that our sample JS app works with some of your quoted UI5 versions)

The app appears to be working in all of these version, 

The wdio.conf file, was generated automatically and includes this line in the config: 
`    services: ['chromedriver','ui5'],`

The fact that wdi tests work which do not use selector might suggest that it is installed properly?

**Comment by vobu** [comment]
> > and just to doublecheck: the app itself works with the changed UI5 version(s)? and you have the `ui5` service configured in `wdio.conf.(j|t)s` (https://ui5-community.github.io/wdi5/#/installation?id=c-manual-installation)? Because when the tests error out with all specified UI5 versions, a possible cause could be that `wdi5` itself (via its' technical name `wdio-ui5-service`) is not loaded at test-time. (Plus I just checked that our sample JS app works with some of your quoted UI5 versions)
> 
> The app appears to be working in all of these version,

👍 thanks for checking

> The wdio.conf file, was generated automatically and includes this line in the config: ` services: ['chromedriver','ui5'],`

this looks fine as well.
and again just to doublecheck: you're using this conf file when you trigger a run, e.g. via `wdio run your-wdi5-enabled.conf.js`?

> The fact that wdi tests work which do not use selector might suggest that it is installed properly?

actually the opposite is true - when `wdio` works w/o `wdi5`-specific selectors, then this is an indication that `wdi5` `(wdio-ui5-service)` wasn't installed properly!
and as always (and even it may sound obvious), you need to have the app running prior to executing the tests (as described in https://ui5-community.github.io/wdi5/#/running)

**Comment by vobu** [comment]
ahhh, now I think I know what's happening here:
you're bootstrapping the app from the locally installed UI5 npm libs - as denoted by your manifest.json, these don't contain the necessary libs for the `RecordReplay` API we heavily depend on (just as OPA5 does).
Could you for test's sake try bootstrapping the app from the CDN, e.g. `https://ui5.sap.com/resources/sap-ui-core.js`?

**Comment by MichaelSchmid-AGI** [comment]
> > > and just to doublecheck: the app itself works with the changed UI5 version(s)? and you have the `ui5` service configured in `wdio.conf.(j|t)s` (https://ui5-community.github.io/wdi5/#/installation?id=c-manual-installation)? Because when the tests error out with all specified UI5 versions, a possible cause could be that `wdi5` itself (via its' technical name `wdio-ui5-service`) is not loaded at test-time. (Plus I just checked that our sample JS app works with some of your quoted UI5 versions)
> > 
> > 
> > The app appears to be working in all of these version,
> 
> 👍 thanks for checking
> 
> > The wdio.conf file, was generated automatically and includes this line in the config: ` services: ['chromedriver','ui5'],`
> 
> this looks fine as well. and again just to doublecheck: you're using this conf file when you trigger a run, e.g. via `wdio run your-wdi5-enabled.conf.js`?
> 
> > The fact that wdi tests work which do not use selector might suggest that it is installed properly?
> 
> actually the opposite is true - when `wdio` works w/o `wdi5`-specific selectors, then this is an indication that `wdi5` `(wdio-ui5-service)` wasn't installed properly! and as always (and even it may sound obvious), you need to have the app running prior to executing the tests (as described in https://ui5-community.github.io/wdi5/#/running)

when executing wdio we always had the app  and running. 



> ahhh, now I think I know what's happening here: you're bootstrapping the app from the locally installed UI5 npm libs - as denoted by your manifest.json, these don't contain the necessary libs for the `RecordReplay` API we heavily depend on (just as OPA5 does). Could you for test's sake try bootstrapping the app from the CDN, e.g. `https://ui5.sap.com/resources/sap-ui-core.js`?

Seems Unlikely:
Our app is generated with the sap ui5 Plugins
The URL is just put together in three different files In the ui5 .yaml we define the first part of the url :
![image](https://user-images.githubusercontent.com/111853408/200359874-7bbcc821-26db-41e9-85f1-bebdeb1953ae.png)
In the Index,html  we define the second part of the url("resources/sap-ui-core.js") 
The version is set in the manifest.json

I still tested it for you with 1.103.0, 1.104.1 and 1.108.0 and the error still occurs:
![image](https://user-images.githubusercontent.com/111853408/200360810-7ac81439-f8d1-4485-bd01-aef0486dcb29.png)

my colleagues and i are trying to get this working since about 2 Work-days. And  we really would like to finally get this working

**Comment by the-docta** [comment]
@MichaelSchmid-AGI you have the src attribute used twice in bootstrapping. might that be the problem?

![image](https://user-images.githubusercontent.com/65647431/200374691-16ff7444-3ee3-41ed-a99c-926714c77cc4.png)

**Comment by vobu** [comment]
> @MichaelSchmid-AGI you have the `src` attribute used twice in bootstrapping. might that be the problem?
> 
> ![image](https://user-images.githubusercontent.com/65647431/200374691-16ff7444-3ee3-41ed-a99c-926714c77cc4.png)

nice catch!
try cleaning that up to only contain the CDN url (and remove the second superfluous `id` as well) and see whether tests work then.

if not, shoot me an email and we can take things up to enterprise support level to avoid making too much noise here 🤭

**Comment by soraY27** [comment]
Brief Update:
We have solved the problem by implementing a workaround for logging in.
We will look into the new authentication feature (https://github.com/ui5-community/wdi5/tree/feat/authenticator) as soon as it is released. We are looking forward to your good work.
And thanks for the quick replies :)

**Comment by Siolto** [comment]
Hi @xhton,

thanks for providing a public repo to reproduce your issue. I found the problem and it has to do with your usage of the `goTo` function in the `before` hook:

```javascript
before(async () => {
    await wdi5.goTo("/");
  });
```

by simply adapting this to:
```javascript
before(async () => {
    await wdi5.goTo("#");
  });
```
or completely removing it (as it is not necessary for your simple use case) the tests run 🟢. Of course we should improve the error message when the `goTo` function is not used the right way.

Best regards,
Simon
